src/main/java/fi/aalto/cs/apluscourses/intellij/model/ExercisesUpdater.java

```diff
@@ -138,7 +144,9 @@ private void addExercises(@NotNull List<ExerciseGroup> exerciseGroups,
           }
           var exercise = dataSource.getExercise(exerciseId, points, course.getOptionalCategories(),
               course.getTutorials(), authentication, CachePreferences.GET_MAX_ONE_WEEK_OLD, selectedLanguage);
-          exerciseGroup.addExercise(exercise);
+          if (!hiddenElements.shouldHideObject(exercise.getId(), exercise.getName(), selectedLanguage)) {
```

Comment:
Once again, I'm a bit confused of this `ExercisesUpdater` class. We did the filtering already when we loaded dummy exercises, right (in `ExerciseGroup::fromJsonObject`)? But now we do filtering again, because the results of the first round of filtering are not available here. Is it so? :thinking:

I feel a code smell here, but it might be this class in general, not the addition of this PR.

Reply:
It's because the exercise list is initially filled with dummy exercises based on the "exercises" API `/api/v2/courses/266/exercises/`. We need to filter these so that we don't add dummy exercises that are supposed to be hidden.

Later, when the dummy exercises are converted to real exercises, `addExercise` is called, which will remove the dummy exercise and add the new one, so we need to prevent this function call for hidden exercises. The list of all exercise IDs here is obtained from a different API (the `points` one), so we can't just reuse the previous results.

There are two alternative solutions I can think of:
1. During the initial API call that is used to create dummy exercises, do the filtering as it currently is. Save all filtered exercise IDs somewhere, and when converting dummy entries to real ones, instead of calling `shouldHideObject` twice, just check for the exercise ID before even fetching it.
Advantages: no need to fetch hidden exercises
Disadvantages: the code becomes even more complicated, because now we have to store hidden exercise IDs somewhere. Additionally, the ideal place of creating such a list of hidden IDs (`fromJsonObject`) is objectively bad (modifying class fields in a deserialization function isn't good), so creating that list itself will be even more code, and that kinda defeats the purpose

2. Assuming that all dummy exercises are present before we fetch the real ones, we can change `addExercise` to do nothing if there is no dummy exercise in the list.
Advantage: no need to call `shouldHideObject` twice
Disadvantages: I'm not confident in that assumption, this sounds almost like a hack and certainly won't make the code more readable

Personally, I don't think the alternative solutions are better than the current one.

Reply:
Thanks for the explanation! And good reasoning; I agree with you: let's do as it is in this PR.